File: TGC/Implementacoes/I2/test2/DepthFirstSearch.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Arrays;
import java.util.Stack;

public class DepthFirstSearch {
  private int time; // Variável de tempo global
  private boolean[] visited; // Array para marcar os vértices visitados
  private int[] discoveryTime; // Armazena o tempo de descoberta (TD)
  private int[] finishTime; // Armazena o tempo de término (TT)
  private int[] parent; // Armazena o pai (ancestral direto) de cada vértice

  // Construtor da classe
  public DepthFirstSearch(int numVertices) {
    this.time = 0;
    this.visited = new boolean[numVertices + 1];
    this.discoveryTime = new int[numVertices + 1];
    this.finishTime = new int[numVertices + 1];
    this.parent = new int[numVertices + 1];

    // Inicializa os arrays
    Arrays.fill(discoveryTime, 0);
    Arrays.fill(finishTime, 0);
    Arrays.fill(parent, -1); // -1 indica que não tem pai
  }

  // Método para executar a busca em profundidade iterativa
  public void depthFirstSearch(MainGraph graph) {
    for (int v = 1; v <= graph.numVertices; v++) {
      if (discoveryTime[v] == 0) { // Se o vértice ainda não foi visitado
        depthFirstSearchIterative(graph, v); // Executa a busca a partir dele
      }
    }
  }

  // Método que visita um vértice e seus adjacentes de forma iterativa
  private void depthFirstSearchIterative(MainGraph graph, int startVertex) {
    Stack<Integer> stack = new Stack<>();
    stack.push(startVertex);

    while (!stack.isEmpty()) {
      int v = stack.peek();

      if (discoveryTime[v] == 0) { // Se o vértice ainda não foi descoberto
        time++;
        discoveryTime[v] = time;
        visited[v] = true;
      }

      boolean finished = true;

      // Processa os adjacentes de v em ordem lexicográfica
      for (int w : graph.getAdjacents(v)) {
        if (discoveryTime[w] == 0) { // Se w ainda não foi descoberto
          stack.push(w); // Visita o vértice adjacente w
          parent[w] = v; // Define v como pai de w
          System.out.println("Aresta (" + v + ", " + w + ") classificada como TREE");
          finished = false; // Continua a busca a partir de w
          break;
        } else if (finishTime[w] == 0) { // Se w foi descoberto, mas não finalizado
          System.out.println("Aresta (" + v + ", " + w + ") classificada como BACK");
        } else if (discoveryTime[v] < discoveryTime[w]) {
          System.out.println("Aresta (" + v + ", " + w + ") classificada como FORWARD");
        } else {
          System.out.println("Aresta (" + v + ", " + w + ") classificada como CROSS");
        }
      }

      // Se todos os adjacentes de v foram processados, finaliza v
      if (finished) {
        time++;
        finishTime[v] = time;
        stack.pop();
      }
    }
  }

  // Método para salvar os tempos de descoberta e término em arquivo
  public void saveTimes(String filePath) {
    try (BufferedWriter writer = new BufferedWriter(new FileWriter(filePath))) {
      for (int i = 1; i < discoveryTime.length; i++) {
        writer.write("Vértice: " + i + " TD: " + discoveryTime[i] + " TT: " + finishTime[i]);
        writer.newLine();
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
  }
}

File: TGC/Implementacoes/I2/testcpp/DepthFirstSearch.cpp
#pragma once  // Impede a inclusão múltipla do arquivo

#include <iostream>
#include <vector>
#include <fstream>
#include "MainGraph.cpp"  // Inclui a definição da classe MainGraph

class DepthFirstSearch {
public:
    std::vector<bool> visited;
    std::vector<int> discoveryTime;
    std::vector<int> finishTime;
    std::vector<int> parent;
    int time;
    std::vector<std::string> edgeClassifications; // Para armazenar as classificações de arestas

    DepthFirstSearch(int numVertices) {
        visited.resize(numVertices + 1, false);
        discoveryTime.resize(numVertices + 1, 0);
        finishTime.resize(numVertices + 1, 0);
        parent.resize(numVertices + 1, -1);
        time = 0;
    }

    void depthFirstSearch(MainGraph &graph) {
        for (int v = 1; v <= graph.numVertices; v++) {
            if (!visited[v]) {
                depthFirstSearchVisit(graph, v);
            }
        }
        printf("Fim da busca em profundidade\n");
    }

    void depthFirstSearchVisit(MainGraph &graph, int v) {
        time++;
        discoveryTime[v] = time;
        visited[v] = true;

        for (int w : graph.getAdjacents(v)) {
            if (!visited[w]) {
                parent[w] = v;
                std::string edgeInfo = "Aresta (" + std::to_string(v) + ", " + std::to_string(w) + ") classificada como TREE";
                std::cout << edgeInfo << std::endl;
                edgeClassifications.push_back(edgeInfo);
                depthFirstSearchVisit(graph, w);
            } else if (finishTime[w] == 0) {
                std::string edgeInfo = "Aresta (" + std::to_string(v) + ", " + std::to_string(w) + ") classificada como BACK";
                std::cout << edgeInfo << std::endl;
                edgeClassifications.push_back(edgeInfo);
            } else if (discoveryTime[v] < discoveryTime[w]) {
                std::string edgeInfo = "Aresta (" + std::to_string(v) + ", " + std::to_string(w) + ") classificada como FORWARD";
                std::cout << edgeInfo << std::endl;
                edgeClassifications.push_back(edgeInfo);
            } else {
                std::string edgeInfo = "Aresta (" + std::to_string(v) + ", " + std::to_string(w) + ") classificada como CROSS";
                std::cout << edgeInfo << std::endl;
                edgeClassifications.push_back(edgeInfo);
            }
        }

        time++;
        finishTime[v] = time;
    }

    void printTimes() const {
        for (size_t i = 1; i < discoveryTime.size(); i++) {
            std::cout << "Vértice: " << i << " TD: " << discoveryTime[i] << " TT: " << finishTime[i] << "\n";
        }
    }

    // Função para salvar as saídas em um arquivo
    void saveToFile(const std::string &fileName) const {
        std::ofstream outFile(fileName); // Cria ou sobrescreve o arquivo

        if (!outFile.is_open()) {
            std::cerr << "Erro ao abrir o arquivo para escrita." << std::endl;
            return;
        }

        // Salvar as classificações de arestas
        for (const auto &edgeInfo : edgeClassifications) {
            outFile << edgeInfo << std::endl;
        }

        // Salvar os tempos de descoberta e término
        outFile << "\nTempos de descoberta e término:\n";
        for (size_t i = 1; i < discoveryTime.size(); i++) {
            outFile << "Vértice: " << i << " TD: " << discoveryTime[i] << " TT: " << finishTime[i] << "\n";
        }

        outFile.close();
        std::cout << "Resultados salvos no arquivo: " << fileName << std::endl;
    }
};
